govc/importx/archive.go

```diff
@@ -127,23 +132,109 @@ func (t *TapeArchive) Open(name string) (io.ReadCloser, int64, error) {
 
 type FileArchive struct {
 	path string
+	Downloader
 }
 
 func (t *FileArchive) Open(name string) (io.ReadCloser, int64, error) {
 	fpath := name
 	if name != t.path {
-		fpath = filepath.Join(filepath.Dir(t.path), name)
+		index := strings.LastIndex(t.path, "/")
+		if index != -1 {
+			fpath = t.path[:index] + "/" + name
+		}
 	}
 
-	s, err := os.Stat(fpath)
+	f, err := t.OpenFile(fpath)
 	if err != nil {
 		return nil, 0, err
 	}
 
-	f, err := os.Open(fpath)
+	s, err := f.Stat()
 	if err != nil {
 		return nil, 0, err
 	}
 
 	return f, s.Size(), nil
 }
+
+type Downloader struct {
+	sinkFunc func(string) progress.Sinker
+	nocache  bool
+}
+
+func (d Downloader) OpenFile(fpath string) (*os.File, error) {
+	fpath, _, err := d.Ensure(fpath)
+	if err != nil {
+		return nil, err
+	}
+
+	return os.Open(fpath)
+}
+
+func (d Downloader) Ensure(fpath string) (string, bool, error) {
+	if !isRemotePath(fpath) {
+		return fpath, false, nil
+	}
+
+	cache := cachePath(fpath)
+	_, err := os.Stat(cache)
+	if err == nil && !d.nocache {
+		return cache, true, nil
+	}
+
+	return cache, false, d.download(fpath, cache)
+}
+
+func (d Downloader) download(remote, local string) error {
```

Comment:
Can we re-use `soap.Client.DownloadFile` instead?

Reply:
sure.